### ISSUE TITLE
More Flexibility in for Logging Data

### ISSUE THREAD

**Issue by Queuecumber** [PyTorchLightning/pytorch-lightning#7753]
## 🚀 Feature

I've been trying to figure out how to log my data and I can't make it work in a clean way. I think this could be resolved if the logging procedure was a little more flexible.

I work on problems related to JPEG, when I'm ready to evaluate, I need to take three datasets (or more) and evaluate my model on qualities in `range(10, 101, 10)` (so 10, 20, 30, ... 100). I need to output **separate** evaluation metrics for each dataset for each of these qualities (so 30 metrics total output at the end), however, when I go to view these metrics in whatever logger I'm using, I want the X-axis to be quality (instead of step) and the Y axis to by whatever my metric is. 

This doesn't seem possible to do with the current logging setup, ive tried all kinds of tricks like aggregating the metrics in `test_epoch_end`, setting `trainer.global_step = quality`, the only thing that kind of worked is calling my loggers `log_metrics` manually where I can provide the step number. 

Is there a way to do this and if not I'd like to request the feature. 

Secondly, with so many datasets, having them get spit out named "dataset_idx_0, dataset_idx_1 ..." is not very helpful. I've gotten around this by hardcoding a list of the names that I can look up when I make the logging keys so it comes out like 'dataset/quality/metric' but that's also kind of an ugly solution and I need to keep updating it. It would be nice if the `test_dataloader` function could return a map of `dataset name: dataloader` and then the `dataset_idx` parameter passed to `test_step` could be the string key. 

### Motivation

See above

### Pitch

I think we'd need to work out exactly how the logger could be made more flexible but off the top of my head I should be able to set a step number manually (e.g. `self.log(..., step=quality)` ) and PL should take care of aggregating metrics reported in the same step. Either that or there could be a separate `test_on_dataset_end` or something that gets called whenever a dataloader is exhausted (this is different from the `on_epoch_ends` which would be called when **all** dataloaders are exhausted). 

The second part where dataloaders have string names is pretty straightforward.

### Alternatives

I tried a lot, the only thing that works is bypassing the nice logger abilities of PL and going to log_metrics manually, which I'm pretty sure is a bad idea.

**Comment by tchaton** [comment]
Dear @Queuecumber,

Thanks for sharing your thoughts, it is definitely interesting.

1 . For logging on quality, we will have to think how to handle this as it is pretty special use case.
I think Logging directly using `self.trainer.logger.log_metrics` is definitely the cleanest way for me.
Using TorchMetric Metric, you can do this:

Silly example with Accuracy. You can aggregate in the way you want.
```
from torchmetrics import Accuracy

class MyModel(LightningModule):

    def __init__(self, model):
        self.model = model

    def setup(self, ):
        dm = self.trainer.datamodule
        self.map_dl_idx_to_quality = dm.map_dl_idx_to_quality
        self.map_dl_idx_to_name = dm.map_dl_idx_to_name
        self.num_dataloaders = dm.num_dataloaders
        self.val_acc = [Accuracy(num_classes=dm.num_classes) for _ in range(self.num_dataloaders)]

    def validation_step(self, batch, batch_idx, dataloader_idx):
        ...
        self.val_acc[dataloader_idx])(preds, labels)
    
    def validation_epoch_end(self, outputs):
        for dl_idx, quality in self.map_dl_idx_to_quality.items():
            name = self.map_dl_idx_to_quality[dl_idx]
            quality = self.map_dl_idx_to_name[dl_idx]
            metrics = {f"{name}/{quality}/{val_acc}": self.val_acc[dl_idx]}
            self.trainer.logger.log_metrics(metrics, quality)
            self.val_acc[dl_idx].reset()
``` 

2. For "dataset_idx_0, dataset_idx_1", I agree this isn't the best solution right now. We definitely need to find a better way to handle this.

Best,
T.C

**Comment by Queuecumber** [comment]
for 1) the best way to handle it is with tabular data, since I'm using WandB I can log a separate table for each dataset with columns [quality, metric a, metric b, metric c, ...] then do some fancy plotting to get the X axis to be quality and the Y axis to be metric. The way you showed wont exactly work, while the metrics are correctly separated by quality there isnt a good way to put quality on the X axis of a plot. It is a bit corner-cased although I'm sure there are generalizations that would be useful for other tasks, and having a dataset-level end function would be nice. That said it might make sense to leave this one out of PL in general. What would help is a retooling of the logger API to support {images, videos, tables} natively and forward those calls to the underlying experiment object if it supports them (I think this is in the works?)

For part 2 I think it should be fairly straightforward to support a dict return value from `test_dataloader` (and the like) with arbitrary keys mapping to dataloaders and to pass those keys instead of the integer `dataloader_idx` (or append their str or repl automatically to log entries). 

If this sound alright to you I'm happy to PR this as I say on all my issues, but I need to have some feedback on said hypothetical PR from the team so that it gets merged in a timely manner. Lack of engagement on issues/PRs from the team has been a major pain point for me in working with PL thus far.

**Comment by carmocca** [comment]
> (1)

I agree the ties to `global_step` when logging can be limiting for several use cases. Eventually we should be more flexible with what axis is what.

> Secondly, with so many datasets, having them get spit out named ...

Note that you could just not pass multiple datasets and instead call `trainer.{validate,test}(dataloader)` for each dataset.

> Lack of engagement on issues/PRs from the team has been a major pain point for me in working with PL thus far.

Sorry for this. Time is limited and there are many open fronts that have starved our attention on other matters. We'll try to do better 😃

**Comment by Queuecumber** [comment]
> Note that you could just not pass multiple datasets and instead call trainer.{validate,test}(dataloader) for each dataset.

This is worth investigating on my end, although it kind of breaks the nice automated flow you get from using datamodules and having fit do your validation periodically

**Comment by carmocca** [comment]
Just to clarify, you are interested in doing this in the validation part of fitting, not when running validation separately, correct?

**Comment by Queuecumber** [comment]
My use case is actually for `test`, but I think it would generally be useful support something other than an integer index for dataloaders

**Comment by carmocca** [comment]
cc @justusschock @ananthsub for opinions on the dataloader idx part

**Comment by Queuecumber** [comment]
This is kind of what I was talking about, its been almost a week since you pinged those guys for input and we've got nothing

**Comment by edenlightning** [comment]
Thanks for pinging us @Queuecumber! We are trying our best to answer all issues as fast as possible.

**Comment by justusschock** [comment]
@Queuecumber Sorry for the delay!

Regarding your questions:

1.) I think we could go the way with the mapping. "Old-Style" Sequences could most likely be converted to mappings by just having their index as key. The only "issue" would then be the transition from ints (now) to strings (ideally all keys would be strings at some point, even the ones created by sequence-enumeration).

I just wouldn't like to automatically add the name to the key of the logged entity. You would have the option to do so manually, but doing that by default would also limit some users in another way.

2.) I definitely agree that we should provide an optional `step` argument here, which defaults to the global_step if not provided. However, that would then also mean that you have to provide that `step` by yourself (based on the updated dataloader names). The only blocker I see for this: How do you aggregate? The global step is only increasing, so for each step, there is only one logged value (and we don't have to keep them in memory once the current epoch is over). When you can provide very custom steps, this means that you can log to the same step multiple times? What should happen in that case? Should we just overwrite it? Accumulation would be difficult since we cannot keep every value in memory that was ever logged.

**Comment by Queuecumber** [comment]
> @Queuecumber Sorry for the delay!
> 
> Regarding your questions:
> 
> 1.) I think we could go the way with the mapping. "Old-Style" Sequences could most likely be converted to mappings by just having their index as key. The only "issue" would then be the transition from ints (now) to strings (ideally all keys would be strings at some point, even the ones created by sequence-enumeration).

I think this works and shouldn't be too hard to implement. I was actually thinking keys could be anything suitable for a dict key and it's up to the user what to do with the keys.

> I just wouldn't like to automatically add the name to the key of the logged entity. You would have the option to do so manually, but doing that by default would also limit some users in another way.

I think we could keep the default behavior of appending the key as long as it has a str or repr we can use, I agree it needs to be handled carefully though so it doesn't surprise people
 
> 2.) I definitely agree that we should provide an optional `step` argument here, which defaults to the global_step if not provided. However, that would then also mean that you have to provide that `step` by yourself (based on the updated dataloader names). The only blocker I see for this: How do you aggregate? The global step is only increasing, so for each step, there is only one logged value (and we don't have to keep them in memory once the current epoch is over). When you can provide very custom steps, this means that you can log to the same step multiple times? What should happen in that case? Should we just overwrite it? Accumulation would be difficult since we cannot keep every value in memory that was ever logged.

This is a tricky one, and probably it needs to be thought out very carefully before we proceed to a PR. For my case, I would want the values averaged for every step, but I can see sum, replace, and custom accumulator all being useful, which all have different overhead associated with them.

**Comment by justusschock** [comment]
> I think this works and shouldn't be too hard to implement. I was actually thinking keys could be anything suitable for a dict key and it's up to the user what to do with the keys.

Do you want to give the PR a shot? That'd be great! Please ping me directly on it and I promise I won't take that long to reply this time :)

> I think we could keep the default behavior of appending the key as long as it has a str or repr we can use, I agree it needs to be handled carefully though so it doesn't surprise people

So you always want to convert it to a string and then treat it as we currently do? I'm fine with that (we just need to provide a meaningful error when it cannot be converted)

> This is a tricky one, and probably it needs to be thought out very carefully before we proceed to a PR. For my case, I would want the values averaged for every step, but I can see sum, replace, and custom accumulator all being useful, which all have different overhead associated with them.

Yes. i think the main question for me is here: When do you want the accumulation to happen? Currently the only option is to accumulate at the end of the epoch (afaik). My main concern is how we avoid blowing up the memory too much. After we figured that out, the different reductions should actually be no-brainers

**Comment by Queuecumber** [comment]
I should be able to get to this PR this week, will ping you when I send it in

**Comment by Queuecumber** [comment]
Still on my TODO list, sorry for the delay
